### PR TITLE
Ability to add custom facet in OpenLineage events

### PR DESCRIPTION
**Pull request by anandhimurali**
**Motivation**
Currently, users need to write custom extractor inorder to add custom facets in the lineage events. Instead, providing a way to inject custom facets by accepting a config.

**Modifications**
The `get_custom_facets` function in the OpenLineage provider is enhanced to fetch the Airflow config `AIRFLOW__OPENLINEAGE__CUSTOM_RUN_FACETS`, execute those functions and append their return values to the run facet of the lineage event for all the operators.

**Result**
Enables users to add custom facets to the lineage events by defining functions instead of custom extractor.

**Configs documentation for `AIRFLOW__OPENLINEAGE__CUSTOM_RUN_FACETS`**

image

**User Docs**

image

image

image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/openlineage/provider.yaml
Comment (eladkal) [inline airflow/providers/openlineage/provider.yaml]:
```suggestion
        version_added: 1.8.0
```
We can assume this will be added in next feature release

**Comment by XD-DENG** [comment]
Hi @eladkal , @anandhimurali is my colleague from the same team and I'm really excited about her 1st PR to Airflow!

Do you have an idea who among the committers may be the best to help review this openlineage-related change?

Any suggestion/input would be greatly appreciated. Thanks a lot!

**Comment by eladkal** [comment]
Ping @mobuchowski for review :)

**Comment by mobuchowski** [comment]
Hey @anandhimurali @XD-DENG I think ability to add custom facets to arbitrary operators is a great idea.

I have two points I think it's worth addressing: 

1. Whether we pass all the necessary info. I think given task_instance we can call `get_template_context` and get all the data, so this might not be an issue.
2. Whether this acts "per DAG/Task" or it's a mechanism that works for all the tasks equally - this might be defined inside the function so it still might not be an issue other than complicating the function a bit more.

Overall I like the solution, but might be good to specify the goals regarding 1) and 2) - because for sure, later, someone will come and ask "how to make this custom facet work only on some particular tasks" and "how do I pass some custom info to my function".

File: airflow/providers/openlineage/conf.py
Comment (kacpermuda) [inline airflow/providers/openlineage/conf.py]: I don't think the `OPENLINEAGE_CUSTOM_FACET_FUNCTIONS` variable is necessary, we can stick to Airflow config. We use `OPENLINEAGE_` variables in some other configs as they are legacy solutions that still need to be supported. For new features we can omit them.

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [inline airflow/providers/openlineage/utils/utils.py]: I would add a check for duplicates here (check if we are overwriting some facet already and maybe put some information in logs) and possibly also disallow some predefined facet names (keys in dict), so that some base facets are not overwritten by user accidentally. Or maybe we should allow overwriting and give the user full control? WDYT @mobuchowski @JDarDagran

File: airflow/providers/openlineage/provider.yaml
Comment (kacpermuda) [inline airflow/providers/openlineage/provider.yaml]:
```suggestion
          Register custom facet functions, that return facets to append as run facets, by passing a string of semicolon separated full import paths.
```

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [inline airflow/providers/openlineage/utils/utils.py]: `func` can also be None, so we should maybe include that in type hint?

File: airflow/providers/openlineage/provider.yaml
Comment (kacpermuda) [reply airflow/providers/openlineage/provider.yaml]: Not sure if we should include that here as it's an implementation detail. If we change the input parameters it may be easy to forget to change that in the yaml, but maybe I'm wrong?
> The functions should accept a TaskInstance input argument and return a dictionary.

**Review by kacpermuda — COMMENTED** [review]
I like the idea 🚀  Apart from what @mobuchowski mentioned above, I added some small comments about the code itself.

File: airflow/providers/openlineage/conf.py
Comment (mobuchowski) [reply airflow/providers/openlineage/conf.py]: I think good question is whether we can do it without global configs? I think this might be good to be configurable on task/dag/global level, rather than fixed for whole Airflow instance. It goes well with our effort to have lineage be incrementally enabled.

**Comment by kacpermuda** [comment]
Hey @anandhimurali @XD-DENG, is there anything i could help you with here? I really like the idea and would like to see it happen, let me know 😄

**Comment by anandhimurali** [comment]
>Hey @anandhimurali @XD-DENG, is there anything i could help you with here? I really like the idea and would like to see it happen, let me know 😄

Hi @kacpermuda, Thanks for the interest and I'm excited to hear that. Sorry I've been busy for the last few weeks. I'll look into the comments this week.

File: airflow/providers/openlineage/utils/utils.py
Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: Addressed in a3014683241b130ad9729dc462efede1160361e4

File: airflow/providers/openlineage/utils/utils.py
Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: At the [place where the run facets are merged](https://github.com/apache/airflow/blob/9ea78d9d726d9ddb6109a7fba0c3a838f8a05610/airflow/providers/openlineage/plugins/listener.py#L147), the results from `get_custom_facets` is unwrapped before the `get_airflow_run_facet`. So user can't overwrite the Airflow base facets.

Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: Working on adding duplicate check on the custom facet keys and ignoring such duplicate keys by adding logging.

File: airflow/providers/openlineage/conf.py
Comment (anandhimurali) [reply airflow/providers/openlineage/conf.py]: >I don't think the OPENLINEAGE_CUSTOM_FACET_FUNCTIONS variable is necessary, we can stick to Airflow config. We use OPENLINEAGE_ variables in some other configs as they are legacy solutions that still need to be supported. For new features we can omit them.

Addressed in ec330478bbefc00118effe6ec9f4930907e3d273

Comment (anandhimurali) [reply airflow/providers/openlineage/conf.py]: >I think good question is whether we can do it without global configs? I think this might be good to be configurable on task/dag/global level, rather than fixed for whole Airflow instance. It goes well with our effort to have lineage be incrementally enabled.

@mobuchowski , User can choose to `return None` in the custom facet function, if they don't want to add the custom facet for certain tasks/dags. Thinking if the complexity is going to increase if we handle the task/dag/global level settings at config level.

File: airflow/providers/openlineage/utils/utils.py
Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: >Working on adding duplicate check on the custom facet keys and ignoring such duplicate keys by adding logging.

Addressed in 03ed172b6ba140eaad012872c5e0939c16a0e7da

**Comment by anandhimurali** [comment]
> I have two points I think it's worth addressing:
> 1. Whether we pass all the necessary info. I think given task_instance we can call get_template_context and get all the data, so this might not be an issue.
> 2. Whether this acts "per DAG/Task" or it's a mechanism that works for all the tasks equally - this might be defined inside the function so it still might not be an issue other than complicating the function a bit more.
>
> Overall I like the solution, but might be good to specify the goals regarding 1) and 2) - because for sure, later, someone will come and ask "how to make this custom facet work only on some particular tasks" and "how do I pass some custom info to my function".

Hi @mobuchowski, 
1. Yes, I think `task_instance` is sufficient to fetch necessary info about the run context.
2. The `custom_facet_functions` defined works equally for all tasks. Yes, user can choose to `return None` for scenarios where they won't to add the custom facet.

Let me know if this needs to be documented anywhere. Thanks.

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [inline airflow/providers/openlineage/utils/utils.py]:
```suggestion
        if not func:
            log.warning("OpenLineage is unable to import custom facet function `%s`; will ignore it.", custom_facet_func)
            continue
        facets: dict[str, BaseFacet] = func(task_instance)
```

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [inline airflow/providers/openlineage/utils/utils.py]:
```suggestion
            duplicate_facet_keys = [facet_key for facet_key in facet.keys() if facet_key in custom_facets]
            if duplicate_facet_keys:
                log.warning(
                    "Duplicate OpenLineage custom facets key(s) found: `%s` from function `%s`.",
                    ", ".join(duplicate_facet_keys),
                    custom_facet_func,
                )
            log.debug("Adding OpenLineage custom facet with key(s): `%s` from function `%s`.", tuple(facet), custom_facet_func)
            custom_facets.update(facet)
```

**Review by kacpermuda — COMMENTED** [review]
As you already mentioned, i think this should be documented for the users. We should include the function signature with type hints, so that users know how to properly implement it + some examples from our side. 

Probably [here](https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-openlineage/guides/user.rst) and [here](https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-openlineage/guides/developer.rst)

**Comment by mobuchowski** [comment]
@anandhimurali any interest in getting this PR into shape and approved? I think this is good, really useful feature.

**Comment by anandhimurali** [comment]
> @anandhimurali any interest in getting this PR into shape and approved? I think this is good, really useful feature.

@mobuchowski Yes, definitely. I just have to add some documentation and will get this done this week. Thanks and sorry for the delay.

**Comment by ahidalgob** [comment]
I think it would be good to rename this functionality to `custom_run_facets`, I think we might need to inject other facets too, specifically `Job.facets`.

**Comment by anandhimurali** [comment]
>I think it would be good to rename this functionality to custom_run_facets, we might need to inject other facets too, specifically Job.facets.

Addressed in 3a11b7625c896252c681bf1f217f0cce1f5eca57.

**Comment by anandhimurali** [comment]
Getting some documentation error on `breeze build-docs openlineage`. Trying to debug and fix this.

image

**Comment by potiuk** [comment]
> Getting some documentation error on `breeze build-docs openlineage`. Trying to debug and fix this.
> 
> image

FYI. Sometimes `--clean-build` flag helps.

File: airflow/providers/openlineage/provider.yaml
Comment (kacpermuda) [inline airflow/providers/openlineage/provider.yaml]: This will probably be 1.10.0 as 1.9.0 should be released before this gets merged.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst
Comment (kacpermuda) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst]:
```suggestion
    def get_my_custom_facet(task_instance: TaskInstance):
        operator_name = task_instance.task.operator_name
        if operator_name == "BashOperator":
            return None
        job_unique_name = f"TEST.{task_instance.dag_id}.{task_instance.task_id}"
```

Including some simple example here can make it easier for the users to add conditional facets.

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [inline airflow/providers/openlineage/utils/utils.py]: It's more general comment but currently `get_custom_facets()` function is called only once, after task instance has started. 

I believe we need to decide whether custom run facets should receive the task before or after execution. Users might want to use elements like `query_id` generated during execution in their custom facets. While I don't have a strong preference either way, it's important to note this decision consciously in the documentation.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst
Comment (kacpermuda) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: If we decide that this function will receive the task after the execution, we can also show in the example how to retrieve some attribute from the operator.

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [inline airflow/providers/openlineage/utils/utils.py]: I noticed that in examples here we are using attrs.asdict(), so we are providing the facets as dicts, but here we actually expect type[BaseFacet]. Could you adjust the typing and examples so that it matches?

Comment (kacpermuda) [reply airflow/providers/openlineage/utils/utils.py]: And one more question / decision to make: should we wrap the function call with try/except to avoid crashing the whole event emission or do we let the user crash it if they provided the function that breaks? I would vote for putting a try/except with some descriptive logging here.

**Review by kacpermuda — COMMENTED** [review]
Apologies for reaching out again 😄. I wanted to say that I really appreciate the documentation you wrote, and I’m a big fan of the entire feature. I’m not trying to find faults or be difficult, but I had two additional thoughts. I’ve included them in the comments along with two minor improvement suggestions.

File: airflow/providers/openlineage/provider.yaml
Comment (anandhimurali) [reply airflow/providers/openlineage/provider.yaml]: Addressed in c1bc84efc2a454dc5eb4c0e33eb206506b06e445.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst
Comment (anandhimurali) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: >Including some simple example here can make it easier for the users to add conditional facets.

Addressed in 3455b732e64a70fc3c041390d8f841987c47828f.

File: airflow/providers/openlineage/utils/utils.py
Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: >And one more question / decision to make: should we wrap the function call with try/except to avoid crashing the whole event emission or do we let the user crash it if they provided the function that breaks? I would vote for putting a try/except with some descriptive logging here.

Makes sense. Added try-except in db25be02a92c2f0f81b27710d92fb760365f2148.

Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: >I noticed that in examples here we are using attrs.asdict(), so we are providing the facets as dicts, but here we actually expect type[BaseFacet]. Could you adjust the typing and examples so that it matches?

We expect that the functions return `dict` or `None`. Addressed the typing error in 3790b7410a667b2cfe691b8a775d3ac0e31c6fc8.

File: airflow/providers/openlineage/utils/utils.py
Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: >It's more general comment but currently get_custom_facets() function is called only once, after task instance has started.

Updated documentation to capture this, 426cc856fd9ec4071df4fcd50138c7f6c2b2252c.

**Comment by anandhimurali** [comment]
> FYI. Sometimes --clean-build flag helps.

Thanks @potiuk, `--clean-build` was helpful to overcome those documentation build errors.

File: airflow/providers/openlineage/utils/utils.py
Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: >I believe we need to decide whether custom run facets should receive the task before or after execution. Users might want to use elements like query_id generated during execution in their custom facets. While I don't have a strong preference either way, it's important to note this decision consciously in the documentation.

Since the [get_custom_facets](https://github.com/apache/airflow/blob/9d92feda7fae0f014743153ce69656eb339fc420/airflow/providers/openlineage/plugins/listener.py#L166) is invoked only during the OpenLineage START event generation, only the state at the start of the TaskInstance is available to the custom facet functions. 

However, if the state after task execution needs to be made available to the functions, it can be only be done during the COMPLETE or other adhoc lineage events? IMO, the scope of changes will be larger for this and can be taken as a follow-up PR. Let me know what you think.

File: airflow/providers/openlineage/utils/utils.py
Comment (kacpermuda) [reply airflow/providers/openlineage/utils/utils.py]: I was more concerned about the type of the dict's key, so whether the function should produce:
`facet: dict[str, Type[BaseFacet]] | None` or `facet: dict[str, dict] | None` ? Maybe it doesn't matter, but then we should probably also let the users know 😉

Comment (anandhimurali) [reply airflow/providers/openlineage/utils/utils.py]: Return type can be `facet: dict[str, dict] | None`. Addressed in b5d3f37b5aa4171d112823ade8c41bf12743ef8b.

**Comment by anandhimurali** [comment]
>@anandhimurali If you need any help to make the CI pass let me know, If you run the pre-commits locally then all of the static checks will pass. When it comes to 2.7 compatibility tests, you simply need to add start_date to the test DAGs you are creating and it should also work fine.

@kacpermuda, Thanks for letting me know about this. The necessary fixes are addressed in 81a8282b60c5bb28e46b39672ca99707dd7c7454 after the pre-commit checks.

File: tests/providers/openlineage/utils/test_utils.py
Comment (kacpermuda) [inline tests/providers/openlineage/utils/test_utils.py]: @anandhimurali  I believe adding something like this to all places where you create DAG in tests would solve the problem with the last CI check failing and would make it all green😄 
```suggestion
        task=EmptyOperator(task_id="test-task", dag=DAG("test-dag", start_date=datetime.datetime(2024, 7, 1))),
```

Comment (anandhimurali) [reply tests/providers/openlineage/utils/test_utils.py]: Thanks again. Addressed in 8c404b93bab1624e8ec893e1fd6b2ff86682a773.

File: tests/system/providers/google/cloud/datapipelines/resources/__init__.py
Comment (kacpermuda) [inline tests/system/providers/google/cloud/datapipelines/resources/__init__.py:1]: This file looks like some leftover from another branch

**Review by kacpermuda — APPROVED** [review]
LGTM

File: tests/system/providers/google/cloud/datapipelines/resources/__init__.py
Comment (anandhimurali) [reply tests/system/providers/google/cloud/datapipelines/resources/__init__.py:1]: This was added by the pre-commit hook. I tried again now by removing the file and able to get pre-commit checks pass. So removed this file in 18526d7aa317666825d2c737f41d353ef8f5f62e.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst
Comment (JDarDagran) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst:506]: nit: there's no need to translate to dict here, OL client Serde takes care of it

File: tests/providers/openlineage/utils/custom_facet_fixture.py
Comment (JDarDagran) [inline tests/providers/openlineage/utils/custom_facet_fixture.py:47]: nit: same as in docs comment, no need for that

**Review by JDarDagran — APPROVED** [review]
one small thing, otherwise great job! 🚀

**Review by mobuchowski — APPROVED** [review]

**Comment by mobuchowski** [comment]
Thanks @anandhimurali for implementing this feature 🚀

**Comment by MassyB** [comment]
@mobuchowski do you think this is only useful for Airflow ? I would like to have the same configuration for `dbt-ol` as well. 
The signature of the function would be different (an empty args function is enough for my usecase)

**Pull request merged**
